tecton_client/_internal/tecton_client.py

```diff
@@ -0,0 +1,160 @@
+        self._api_key = api_key
+        self._base_url = urljoin(url, "/api/v1/")
+
+        headers = httpx.Headers(
+            {"Authorization": "Tecton-key " + api_key, "User-Agent": "tecton-http-python-client " + tecton_version}
+        )
+        if client is None:
+            self._client = httpx.Client(headers=headers)
+        else:
+            self._client = client
+            # add the headers to the existing client headers
+            self._client.headers.update(headers)
+
+    def get_features(
+        self,
+        *,
+        feature_service_name: Optional[str] = None,
+        feature_service_id: Optional[str] = None,
+        join_key_map: Optional[Dict[str, Union[int, str, type(None)]]] = None,
+        request_context_map: Optional[Dict[str, Any]] = None,
+        metadata_options: Optional[Dict[str, bool]] = None,
+        workspace_name: Optional[str] = None,
+        request_options: Optional[Dict[str, bool]] = None,
+        allow_partial_results: bool = False,
+    ) -> GetFeaturesResponse:
+        self._validate(feature_service_id, feature_service_name, workspace_name)
+        request_data = self._build_get_features_request(
+            feature_service_id,
+            feature_service_name,
+            join_key_map,
+            request_context_map,
+            workspace_name,
+            metadata_options,
+            allow_partial_results,
+            request_options,
+        )
+        resp = self._client.post(url=urljoin(self._base_url, "feature-service/get-features"), json=request_data)
+
+        try:
+            resp.raise_for_status()
+        except HTTPStatusError as exc:
+            raise convert_exception(exc) from exc
+
+        return GetFeaturesResponse.from_dict(json.loads(resp.text))
+
+    def get_feature_service_metadata(
+        self,
+        *,
+        feature_service_name: Optional[str] = None,
+        feature_service_id: Optional[str] = None,
+        workspace_name: Optional[str] = None,
+    ) -> GetFeatureServiceMetadataResponse:
+        self._validate(feature_service_id, feature_service_name, workspace_name)
+        request_data = self._build_get_feature_service_metadata_request(
+            feature_service_id=feature_service_id,
+            feature_service_name=feature_service_name,
+            workspace_name=workspace_name,
+        )
+        resp = self._client.post(url=urljoin(self._base_url, "feature-service/metadata"), json=request_data)
+        try:
+            resp.raise_for_status()
+        except HTTPStatusError as exc:
+            raise convert_exception(exc) from exc
+
+        return GetFeatureServiceMetadataResponse.from_dict(json.loads(resp.text))
+
+    def _build_get_features_request(
+        self,
+        feature_service_id,
+        feature_service_name,
+        join_key_map,
+        request_context_map,
+        workspace_name,
+        metadata_options,
+        allow_partial_results,
+        request_options,
+    ):
```

Comment:
This could be a module-level func rather than a method. Then it could more easily be shared across async/sync impls

Reply:
Generally I imagined we'd have some type of `utils` file which we would share across async/sync it would hold the logic for constructing request/response objects + URLs. And then `TectonClient` and `AsyncTectonClient` would each use these same utilities, but against their resective `HttpClient` and `AsyncHttpClient` objects

Reply:
See if the base-class thing works OK. There is a reference here to 'self' (self.default_workspace_name); I can factor it out of the base-class into a helper method if that's preferred.